DOCS-4588-4.x-distributed-locking-kt (#987)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/distributed-locking.adoc

= Distributed Locking
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: distributed locking, cluster

Mule runtime engine provides the ability to create locks for synchronizing access to resources within Mule components. To manage concurrent access to resources, Mule provides a lock factory that you can access programmatically by scripts or in custom extensions built with the xref:1.1@mule-sdk::getting-started.adoc[Java SDK].

> Any locks you create with Mule's lock factory work seamlessly on deployment models which use a single server or a cluster of servers. This means that if you have a server running the same flow in multiple threads or a cluster environment running the same app, you can guarantee resource synchronization with Mule locks. Additionally, Mule's locking system offers a simple API to access shared locks. Learn how to create locks and sharing them between operations in the following sections.

Reviewer: ```suggestion
Any locks you create with the Mule lock factory work seamlessly on deployment models that use either a single server or a cluster of servers. This means that if you have a server running the same flow in multiple threads, or a cluster environment running one app in multiple locations, you can guarantee resource synchronization with Mule locks. Additionally, the Mule locking system offers a simple API to access shared locks.
```

Reviewer: Edits above:
Avoid making a trademark possessive. 
Please reread to make certain that my suggested rewording about one app in multiple locations is technically accurate.
Remove texts that refer to "this" or "following" text sections, topics, and so on (omit self-reference).

== Creating a Lock

> To create a lock, follow these tasks:

Reviewer: Are these called "tasks" because they are not chronological steps?
If they are not chronological, use an unordered list (bullets) rather than numbers.
```suggestion
To create a lock, perform these tasks:
```

> . Gain access to a `LockFactory` through injection. +

Reviewer: What is a LockFactory? What is "LockFactory" the name or the syntax for? Can you supply a noun object (like "class")?

Reviewer: Or should we be saying "Gain access to a Mule lock factory through injection"?
We need a common noun.

Reviewer: ...and, as above, this should be a bullet rather than Step 1, if this list is not chronological.

> If you are building an extension with the Java SDK and need locking for one of your  operations, you need to inject the `LockFactory` into your class. +

Reviewer: ```suggestion
If you are building an extension with the Java SDK and need locking for one of your operations, you inject `LockFactory` into your class. +
```

> . Use the `LockFactory` to start creating locks. +

Reviewer: ```suggestion
. Use `LockFactory` to start creating locks. +
```
OR
```suggestion
. Use the lock factory to start creating locks. +
```

> Refer to the following sample code which declares an extension operation that manages code access through Mule locks and thus prevents synchronization issues when the operation is executing in parallel (like in a scatter gather scope or in different cluster nodes). +

Reviewer: This text is a little problematic, because it introduces sample code that the user would use to perform this bulleted task, but that code is not shown until after we list other, different tasks.
I also wonder if we really have to reaffirm that the lock prevents sync issues (Line 9 text), and I wonder if it will confuse readers to provide a new scenario (scatter gather), omit the "single flow in multiple threads", and repeat the cluster scenario, without mentioning the single application (also Line 9 text).

Reviewer: Consider...:

Reviewer: ```suggestion
See the following sample code that declares an extension operation that manages code access through Mule locks. +
```

. Use an identifier for the lock you create. +

> The identifier allows you to access the same lock from different threads without having to explicitly share the same lock instance between them. +

Reviewer: ```suggestion
An identifier enables you to access one lock from different threads without having to explicitly share the lock instance between them. +
```
Does "explicitly" = "physically"?

> This lock implements the `java.util.concurrent.Lock` interface, so calling the method `lock` will make the thread wait until it acquires the lock. +

Reviewer: ```suggestion
The following sample code implements the `java.util.concurrent.Lock` interface, so calling the method `lock`v makes the thread wait until it acquires the lock. +
```
Did I interpret this correctly, that "This lock" refers to the lock created in the code sample below?

> After obtaining the lock, the critical code section can be executed and, no matter what the result is, we want to release the obtained lock (to prevent deadlocks on other threads waiting for this lock), so we wrap the `lock.unlock` call in a `finally` block.

Reviewer: This sentence is pretty long and involves several actions, or what sound like actions (I am having some trouble knowing if the critical code is executed and whether it is automatic or user-driven). I've taken a stab at rewording for clarity, but please read it carefully to ensure that it does not change your intended meaning, and feel free to come up with your own wording!
```suggestion
After you create the lock, you must also create a lock release mechanism to prevent deadlocks on other threads waiting for this lock. You accomplish this by enclosing the `lock.unlock` call in a `finally` block:
```

[source,java,linenums]
----
public class ClusterOperations {

  @Inject
  private LockFactory lockFactory;

  public void sharedResourceOperation() {
      Lock lock = lockFactory.createLock("sharedResourceId");
      lock.lock();
      try {
          // Lock acquired, execute critical code.
      } finally {
          lock.unlock();
      }
  }
}
----

> 

Reviewer: Because what follows is a scenario-specific way to create a lock, can we give it a title:
Create a Lock Using Scripting
...and then maybe retitle the previous topic to be specific to the scenario that it documents? 
Would that be Create a Lock Without Scripting, or is there a better way to say this?

> If you are using the scripting extension, you must first access the registry to obtain the `LockFactory` and then create locks through scripting code. The following example uses Groovy to achieve the same result as the previous example: obtain the  `LockFactory` (this time through the `Registry`), create a new `Lock` for the given ID, wait until obtaining the lock to execute critical code and after that release the lock.

Reviewer: Do we need to identify to what the scripting extension is an extension? "...using the scripting extension to <what>...".
```suggestion
If you are using the scripting extension, you must first access the registry to obtain the lock factory and then create locks through scripting code. The following example uses Groovy:
```

[source,groovy,linenums]
----
lockFactory = registry.lookupByName("_muleLockFactory").get()
lock = lockFactory.createLock("sharedResourceId")
lock.lock()
try {
	// Lock acquired, execute critical code.
} finally {
	lock.unlock()
}
----

== Sharing a Lock Between Operations

> Locking is not only useful to create locks for many threads running the same operation, but it can also be used for different operations that may be running in parallel. Using the same lock ID to create a lock in different operations, will make them obtain access to the same lock. The following code sample describes two custom operations which create locks using the same lock ID.

Reviewer: ```suggestion
Locking is useful not only for many threads running the same operation, but also for different operations running in parallel. Using the same lock ID to create a lock in different operations allows those operations to obtain access to the same lock. The following code sample illustrates two custom operations that create locks using the same lock ID:
```

[source,java,linenums]
----
public class ClusterOperations {

  @Inject
  private LockFactory lockFactory;

  public void sharedResourceOperationA() {
      Lock lock = lockFactory.createLock("sharedResourceId");
      lock.lock();
      try {
          // Lock acquired, execute critical code.
      } finally {
          lock.unlock();
      }
  }

  public void sharedResourceOperationB() {
      Lock lock = lockFactory.createLock("sharedResourceId");
      lock.lock();
      try {
          // Lock acquired, execute critical code.
      } finally {
          lock.unlock();
      }
  }
}
----

> 

Reviewer: I wonder if this warning text should be specified at the beginning of the topic set, so that the audience is aware of this issue from the beginning and does not expect to use the `newCondition()` method? Or at the beginning of only the sharing topic?
I think Requirements and Restrictions are sort of like Prerequisites, and that audiences should know about them from the beginning, so they can choose whether or not to read on, if their environment does not support what is going to be documented.

[WARNING]
====

> `LockFactory` creates a `Lock` that implements `java.util.concurrent.Lock` but does not support the method `newCondition()`. Using `newCondition()` causes Mule to throw an `UnsupportedOperationException`. Mule does support other methods of interfacing with a lock, such as `lockInterruptibly()` and `tryLock()`.

Reviewer: ```suggestion
`LockFactory` creates a lock that implements `java.util.concurrent.Lock` but does not support the method `newCondition()`. Mule does support other methods of interfacing with a lock, such as `lockInterruptibly()` and `tryLock()`.
```
